test/test_block_adjacent_difference.cu

```diff
@@ -0,0 +1,521 @@
+    thread_data[item]      = idx < valid_items ? data[idx] : DataType();
+  }
+  __syncthreads();
+
+  if (ReadLeft)
+  {
+    DataType thread_preds[ItemsPerThread];
+
+    BlockAdjacentDifferenceT(temp_storage)
+      .SubtractLeft(thread_result,
+                    thread_data,
+                    thread_preds,
+                    CustomDifference(),
+                    oob_default);
+  }
+  else
+  {
+    BlockAdjacentDifferenceT(temp_storage)
+      .SubtractRight(thread_result,
+                     thread_data,
+                     CustomDifference(),
+                     oob_default);
+  }
+
+  for (unsigned int item = 0; item < ItemsPerThread; item++)
+  {
+    const unsigned int idx = thread_offset + item;
+
+    if (idx >= valid_items)
+      break;
+
+    data[idx] = thread_result[item];
+  }
+}
+
+template <typename DataType,
+          unsigned int ItemsPerThread,
+          unsigned int ThreadsInBlock,
+          bool ReadLeft = false>
+void BlockAdjacentDifferenceLastTileTest(DataType *data, unsigned int valid_items)
+{
+  BlockAdjacentDifferenceLastTileTestKernel<DataType,
+                                            ThreadsInBlock,
+                                            ItemsPerThread,
+                                            ReadLeft>
+    <<<1, ThreadsInBlock>>>(data, valid_items);
+
+  CubDebugExit(cudaPeekAtLastError());
+  CubDebugExit(cudaDeviceSynchronize());
+}
+
+template <typename DataType,
+          unsigned int ItemsPerThread,
+          unsigned int ThreadsInBlock,
+          bool ReadLeft = false>
+void BlockAdjacentDifferenceTest(DataType *data,
+                                 unsigned int valid_items,
+                                 DataType oob_default)
+{
+  BlockAdjacentDifferenceTestKernel<DataType,
+                                    ThreadsInBlock,
+                                    ItemsPerThread,
+                                    ReadLeft>
+    <<<1, ThreadsInBlock>>>(data, valid_items, oob_default);
+
+  CubDebugExit(cudaPeekAtLastError());
+  CubDebugExit(cudaDeviceSynchronize());
+}
+
+template <
+  typename DataType,
+  unsigned int ItemsPerThread,
+  unsigned int ThreadsInBlock>
+void TestLastTile(unsigned int num_items,
+                  thrust::device_vector<DataType> &d_data)
+{
+  thrust::sequence(d_data.begin(), d_data.end());
+
+  constexpr bool read_left = true;
+  constexpr bool read_right = false;
+
+  BlockAdjacentDifferenceLastTileTest<DataType,
+                                      ItemsPerThread,
+                                      ThreadsInBlock,
+                                      read_left>(thrust::raw_pointer_cast(
+                                                   d_data.data()),
+                                                 num_items);
+
+  {
+    const std::size_t expected_count = num_items - 1;
+    const std::size_t actual_count =
+      thrust::count(d_data.begin(), d_data.end(), static_cast<DataType>(1));
```

Comment:
I'm a little nervous about having so many tests producing a uniform output, both for the block and device tests. There are a lot of potential issues that these wouldn't catch (transposed outputs, invalid index calculations, etc). We should add some tests that produce a predictable but non-uniform result.

For instance, we could use an input array where `in[i] = (i * (i - 1)) / 2`. Taking the adjacent difference of that will always produce a sequence:

```
idx:       |  0  1  2  3  4  5  6 ...  N        |
input:     |  0  0  1  3  6 10 15 ...  N(N-1)/2 |
sub_left:  |  0  0  1  2  3  4  5 ...  N-1      |
sub_right: |  0 -1 -2 -3 -4 -5 -6 ...  N(N-1)/2 |
```